cocotb/share/lib/simulator/simulatormodule.cpp

```diff
@@ -219,7 +217,6 @@ int handle_gpi_callback(void *user_data)
             PyErr_Print();
 
             gpi_sim_end();
```

Comment:
This now causes `FATAL: We are calling up again` error because we aren't in simulator context when `gpi_embed_event` is called.

Reply:
I had debated whether I should add the line there or not, but we signal to the users we are shutting down in the shutdown callback, so they can handle shutdown then, we don't need to tell them early.

Reply:
Side question: is this ``FATAL: We are calling up again`` printing some more context now since https://github.com/cocotb/cocotb/pull/1743/files is in, or did I close #1736 prematurely?

Reply:
@cmarqu You get the location where the call happened:
```
     0.00ns ERROR    cocotb.gpi                                gpi_embed.cpp:432  in embed_sim_event                 FATAL: We are calling up again
```

Reply:
I think it could be worded better, though.

Reply:
@ktbarrett The comment here in `handle_gpi_callback` suggests that we don't want to return to Python as there was an Exception and we don't know the state of things. With `embed_sim_event` now being called from `gpi_sim_end`, we ARE returning into Python to notify the Scheduler/RM. We get a separate error if the event call into Python fails, so maybe that's enough in the case where Python is really screwed.

Reply:
The call to `gpi_embed_event`. no longer occurs since I've refactored sim_end.